### PR TITLE
filetransfer: show bitrate stats

### PR DESCRIPTION
**Pull request by fippo**
followup for #481 @samdutton

Bitrate calculation code borrowed from the constraints-and-stats demo (which looks like it needs a cleanup...)

### CODE REVIEW

**Comment by samdutton** [comment]
Nice one! 

I noticed that the bitrate won't display for files that transfer in less than a second (the setInterval interval). 

Reducing the interval to 100ms or less means that the bitrate gets displayed, but only for as long as the file takes to transfer. 

Might be worth displaying bitrate in a separate div, and displaying the average (or something) once the transfer has completed.

Also noticed that sometimes (with a shorter interval at least) bitrate is NaN – not sure why. As a workaround, could check for that.

**Comment by fippo** [comment]
How about average bitrate calculated as file.size / time-it-takes-to-transfer? That also works in Firefox (which doesn't support sent/recv for datachannels currently -- https://bugzilla.mozilla.org/show_bug.cgi?id=1136832)

The NaN might happen when getStats internally did not update yet since the calculation is based on the stats reports timestamp. Since bytesReceived wasn't updated I'm just skipping when that happens.

You made me wonder how hard it would be to rip out the bitrate drawing code webrtc-internals uses. Even though I prefer highcharts for drawing that

**Comment by samdutton** [comment]
>
> How about average bitrate calculated as file.size /
> time-it-takes-to-transfer?


Perfect.

BTW - do you want to add a sentence or two that explains what's happening
in the demo?

 bitrate drawing code webrtc-internals uses


Tempting, though it will make the code a bit more complex.

**Comment by fippo** [comment]
done, WDYT?

If I had time I would write something about the different choices of signaling the file metadata... for small files, in-band signaling makes sense, for larger files you want to use the signaling channel etc. Looks like I need to find a conference that lets me talk about this :-)

Also found a way to display the bitrate for Firefox. Because demos that work in all browsers are the web we want :-)

File: src/content/datachannel/filetransfer/index.html [hunk 1]
```diff
@@ -29,28 +29,46 @@
...
+      <div class="progress">
+        <div class="label">Send progress: </div>
+        <progress id="sendProgress" max="0" value="0"></progress>
+      </div>
+
+      <div class="progress">
+        <div class="label">Receive progress: </div>
+        <progress id="receiveProgress" max="0" value="0"></progress>
+      </div>
+      <div id="bitrate"></div>
```
Comment (samdutton) [inline src/content/datachannel/filetransfer/index.html]: Add a line space above this in the code – and with CSS for the layout :).

**Comment by samdutton** [comment]
Nice.

LGTM aside from nitpick.

Could maybe display the bitrate every 100ms - I only saw it flash up once
with a ~6MB file.

**Comment by fippo** [comment]
done. 100ms looked too aggressive to me. Changed to 500ms and and added additional display steps at 100ms and 300ms for small files :-)

**Comment by juberti** [comment]
This would probably work great with my long-suffering PR for drawing arbitrary stats graphs.

**Comment by fippo** [comment]
http://fippo.github.io/webrtc/src/content/datachannel/filetransfer/ :-)

**Comment by ashumeow** [comment]
@fippo Looks awesome! :+1:

**Comment by samdutton** [comment]
Fantastic! Merged.

**Comment by samdutton** [comment]
[webrtc.github.io/samples/src/content/datachannel/filetransfer](http://webrtc.github.io/samples/src/content/datachannel/filetransfer/)

Duly tweeted, G+ posted, and included in forthcoming blog post about the samples repo.

**Comment by fippo** [comment]
And linked in a comment at https://bloggeek.me/send-file-webrtc-data-api/ -- which I suspect is the reason why people still attempt to use RTP datachannels, ignoring that the advice there is from 2013 :-)

**Pull request merged**